### PR TITLE
Added "ProxySQL-aware" reconnect logic for LHM 

### PR DESCRIPTION
**Pull request by EtienneBerubeShopify**
## What am I trying to achieve
Currently, if there is a loss of connection between LHM and the MySQL server, LHM will retry but with the closed connection. This PR addresses the problem + goes deeper with underlying potential data loss problems when used with Proxy SQL.

## Why?
A reconnection can happen whenever the connection between LHM --> MySQL server or LHM --> ProxySQL is lost. However, in some scenarios involving failovers, the writer will change and with the right amount of replication lag, this can cause data loss. 

However, in cases where the connection is loss and the writer is the same, then it is perfectly safe to reconnect and retry the last SQL statement.  

## How
The solution implemented goes as follows: When an LHM is created, it will save the current writer's hostname (`select @@global.hostname`). Before every chunk insert, it will check that the hostname is the same as the initial one. 
- If it is the same, it will continue
- If it is different, it will abort. 
In the event of a connection loss, LHM will then start the reconnection procedure.
- If it can, it will then check the hostname and ensure that the writer is the same. If it is, then it will retry the SQL statement. - If it can't reconnect. Or if the hostname doesn't match the initial one, then it will abort to avoid data-loss. 

_Note: if the hostname is localhost, then it will check for the server's id (i.e. `select @@global.server_id`)_

#### Complete algorithm:
image

## What changed?
Most of the new logic resides in `lib/lhm/sql_retry.rb`. There are other changes in other files like `chunker.rb` and `chunk_insert.rb`, but they don't add any features. However, since the new logic change the retry logic quite drastically, a lot of the tests had to be adapted. This means that a lot of the tests files had to modified in some way.

## What to check?
- Code cleanliness
- Possible edge cases

### CODE REVIEW

**Comment by xliang6** [comment]
Have we enabled the ProxySQL dedicated hostgroup feature for CloudSQL apps? One path forward is to detect whether we are talking to KateSQL or core, and only reconnect for those two?

File: lib/lhm/sql_retry.rb [hunk 1]
```diff
@@ -15,22 +15,109 @@ module Lhm
...
+        @host_retry_config = default_host_retry_config.dup.merge!(retry_for_host_options)
+        @initial_host = hostname
+        @initial_server_id = server_id
+      end
     end
 
-    def with_retries
+    def with_retries(override_prefix=nil)
+      old_prefix =  @log_prefix
+      @log_prefix = override_prefix unless override_prefix.nil?
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply lib/lhm/sql_retry.rb:35]: Before `chunk_insert` was creating its own instance of `SQLRetry` and the logging prefix would be "chunk insert". But now that there's only one instance of `SQLRetry` per Chunker, there needed to be a way to change the prefix once it's passed to the `chunk_insert`

File: lib/lhm/sql_retry.rb [hunk 2]
```diff
@@ -15,22 +15,109 @@ module Lhm
...
+          raise Lhm::Error.new("Could not reconnected to initial MySQL host. Aborting to avoid data-loss") unless same_host_as_initial?
+        end
+        with_time_check do
+          yield(@connection)
+        end
+      rescue Lhm::Error
+        raise
+      rescue StandardError => e
+        raise e unless error_can_trigger_reconnect?(e)
+        reconnect_with_host_check! if @with_consistent_host
```
Comment (shuhaowu) [inline lib/lhm/sql_retry.rb:48]: Wait. I'm under the impression that we do not want to solve this problem now. I thought we only want to check for host consistency and not worry about reconnect until later?

File: lib/lhm/sql_retry.rb [hunk 3]
```diff
@@ -15,22 +15,109 @@ module Lhm
...
       Retriable.retriable(retry_config) do
-        yield(@connection)
+        if @with_consistent_host
+          raise Lhm::Error.new("Could not reconnected to initial MySQL host. Aborting to avoid data-loss") unless same_host_as_initial?
+        end
+        with_time_check do
+          yield(@connection)
+        end
+      rescue Lhm::Error
+        raise
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply lib/lhm/sql_retry.rb:45]: Because `Lhm::Error < StandardError` and we want different flows for exception handling. Otherwise, every "internal" error used for control flow would trigger a reconnection and it would loop. I tried removing it to simplify the logic, and it broke it.

File: lib/lhm/sql_retry.rb [hunk 4]
```diff
@@ -15,22 +15,109 @@ module Lhm
...
       @retry_config = default_retry_config.dup.merge!(options)
+      if (@with_consistent_host = with_consistent_host)
+        @host_retry_config = default_host_retry_config.dup.merge!(retry_for_host_options)
+        @initial_host = hostname
+        @initial_server_id = server_id
+      end
     end
 
-    def with_retries
+    def with_retries(override_prefix=nil)
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply lib/lhm/sql_retry.rb:33]: Good idea, I'll add some comments / link to the document with the explanation of the algorithm

**Comment by shuhaowu** [comment]
> I'm unsure how reliable this can be since Google could failover CloudSQLs on their own without it being a problem for us, which would change the server id, causing an "Abort".

Conversely, what if they re-use server id like they reuse hostname?

File: lib/lhm/sql_retry.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (shuhaowu) [reply lib/lhm/sql_retry.rb:34]: I would just have the documentation inline unless there's some reason you can't. This way the reader don't need to open yet another browser tab.

File: lib/lhm/sql_retry.rb [hunk 5]
```diff
@@ -15,22 +15,124 @@ module Lhm
...
+        end
+        # The time check will log a warning if an action takes more than 5s (there's a possibility a
+        # failover might've happened after the hostname check above)
+        with_time_check do
+          yield(@connection)
+        end
+      rescue Lhm::Error
+        # Since Lhm::Error < StandardError, it is important to separate the flows, otherwise every
+        # error would trigger a reconnect. In this case, some instances of Lhm::Error are raised in a controlled manner
+        # to trigger retries from ` Retriable.retriable`
```
Comment (shuhaowu) [inline lib/lhm/sql_retry.rb:51]: I had to read this comment a few times and still don't really get it. If some "some instances" of Lhm::Error trigger a controlled retry, why not raise something like `TryReconnect` instead of `LhmError`?

File: lib/lhm/sql_retry.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (shuhaowu) [reply lib/lhm/sql_retry.rb:52]: > I tried removing it to simplify the logic, and it broke it.

What's the second "it"? How did it break?

Also following from my previous comment, can't we use a specific internal error class?

File: lib/lhm/sql_retry.rb (hunk 2 above)
Comment (shuhaowu) [reply lib/lhm/sql_retry.rb:55]: Ah right. The other issue is to reconnect to a different host, which is a big problem. When I wrote the comment I thought we're reconnecting to a different host, which we don't want to solve now.

File: lib/lhm/sql_retry.rb [hunk 6]
```diff
@@ -15,22 +15,124 @@ module Lhm
...
+      log_with_prefix("Lost connection to MySQL, will retry to connect to same host")
+      begin
+        Retriable.retriable(@host_retry_config) do
+          # tries to reconnect. On failure will trigger a retry
+          @connection.reconnect!
+          new_host = hostname
+          if new_host == @initial_host
+            # This is not an actual error, but it needs to trigger the Retriable
+            # from #with_retries to execute the desired logic again
+            raise Lhm::Error.new("LHM successfully reconnected to initial host: #{@initial_host}")
```
Comment (shuhaowu) [inline lib/lhm/sql_retry.rb:108]: If this is not an error, can we just not raise an error? Instead we can raise an internal error..?

**Comment by EtienneBerubeShopify** [comment]
> Conversely, what if they re-use server id like they reuse hostname?

This is a VERY good question. I did run a few failovers myself with a test app and found that they are changing servers for every failover (i.e. they don't "dedicate" a MySQL server to your app). So the server id did change, which mean we can catch "Google-level" failovers, but I didn't find anything online on assessing the true uniqueness of `server_id`s. MySQL does [state](https://dev.mysql.com/doc/refman/5.7/en/replication-options.html) that the number must be between 1 and 2^32, which is big but not insanely big. 

However, the goal is to catch and abort on server_id change. So whenever a failover happens, the `server_id` will change, since I don't think they would failover to the same instance. So even if they do re-use a server_id that was previously used, LHM should catch the difference and abort.

File: README.md [hunk 7]
```diff
@@ -109,6 +109,28 @@ end
 to prevent accidental data loss. After successful or failed LHM migrations, these leftover
 tables must be cleaned up.
 
+### Usage with ProxySQL
+LHM has features to recover from connexion loss that are "ProxySQL-aware". There are multiple ways that connection loss could
+induce data loss. Therefore, if LHM is used with ProxySQL, it will check that the MySQL host is consistent across the schema migrations. 
+Here are the required steps to use LHM with ProxySQL:
+1) Ensure that the ActiveRecord::QueryLogs context is set (requires ActiveRecord >= 7.0.0.aplha2). Ex:
+```ruby
+ActiveRecord::QueryLogs.set_context(maintenance: "lhm") do
```
Comment (tiwilliam) [inline README.md:118]: We could always set this annotation on queries to skip this step?

Comment (tiwilliam) [reply README.md:118]: Yes, prefix all queries coming from this gem.

File: lib/lhm/chunk_insert.rb [hunk 8]
```diff
@@ -2,21 +2,15 @@
 
 module Lhm
   class ChunkInsert
-    def initialize(migration, connection, lowest, highest, options = {})
+    def initialize(migration, sql_retry, lowest, highest, options = {})
```
[the remark this answers is not in the document]
Comment (tiwilliam) [reply lib/lhm/chunk_insert.rb:5]: I think it is in scope of this PR. It would make this code much easier to work with in the future.

**Comment by tiwilliam** [comment]
> I'll probably keep `hostname` as for logging it makes it much more readable than a UUID for observability to debug what happened (or on which host did the migration start).

Isn't it enough to just log the hostname on connect if you need it for debugging?

File: lib/lhm.rb [hunk 9]
```diff
@@ -82,15 +83,26 @@ def cleanup_current_run(run, table_name, options = {})
...
+  #
+  # @param [Hash] connection_options Optional options (defaults to: empty hash)
+  # @option options [Boolean] :reconnect_with_consistent_host
+  #   Active / Deactivate ProxySQL-aware reconnection procedure (default to: false)
+  def connection(connection_options = {})
     @@connection ||=
       begin
         raise 'Please call Lhm.setup' unless defined?(ActiveRecord)
-        ActiveRecord::Base.connection
+        Lhm::Connection.new(connection: ActiveRecord::Base.connection, options: connection_options)
```
Comment (Manan007224) [inline lib/lhm.rb:105]: Nit: You don't need to prefix with main module name here. Something like `Connection` should work instead of `Lhm::Connection`. Just to maintain with the consistency of the code.

File: README.md [hunk 10]
```diff
@@ -109,6 +109,28 @@ end
 to prevent accidental data loss. After successful or failed LHM migrations, these leftover
 tables must be cleaned up.
 
+### Usage with ProxySQL
+LHM has features to recover from connexion loss that are "ProxySQL-aware". There are multiple ways that connection loss could
```
Comment (Manan007224) [inline README.md:113]: spelling error`connexion`.

File: README.md [hunk 11]
```diff
@@ -109,6 +109,28 @@ end
...
+2) Add a `mysql_query_rule`:
+```
+{
+  rule_id = 4
+  active = 1
+  match_pattern = "maintenance:lhm"
+  destination_hostgroup = <MySQL writer's hostgroup>
+}
+```
+
```
Comment (Manan007224) [inline README.md:133]: Could we make this description a bit more clearer. For eg, it is mentioned here that one should add a query rule in proxysql to make Lhm work, but the reason is not described.

File: lib/lhm.rb [hunk 12]
```diff
@@ -82,15 +83,26 @@ def cleanup_current_run(run, table_name, options = {})
     Lhm::Cleanup::Current.new(run, table_name, connection, options).execute
   end
 
-  def setup(connection)
-    @@connection = Lhm::Connection.new(connection: connection)
+  # Setups DB connection
+  #
+  # @param [ActiveRecord::Base] connection ActiveRecord Connection
+  # @param [Hash] connection_options Optional options (defaults to: empty hash)
+  # @option options [Boolean] :reconnect_with_consistent_host
```
Comment (Manan007224) [inline lib/lhm.rb:90]: Would there be a scenario where someone wouldn't want to reconnect with with consistent host ? The reason I am asking here is because we could get rid of the optional variable `reconnect_with_consistent_host` if it is always being set to true.

Comment (EtienneBerubeShopify) [reply lib/lhm.rb:90]: Right now we'll use it as a feature flag, so it is `false` by default (so that people would not inadvertently try to reconnect with ProxySQL without having added the required rules)

File: README.md (hunk 11 above)
Comment (EtienneBerubeShopify) [reply README.md:133]: Good catch! Will do!

File: lib/lhm/chunk_insert.rb [hunk 13]
```diff
@@ -11,7 +12,7 @@ def initialize(migration, connection, lowest, highest, options = {})
     end
 
     def insert_and_return_count_of_rows_created
-      @connection.update(sql, @retry_options)
+      @connection.update(Lhm::ProxySQLHelper.tagged(sql), @retry_options)
```
Comment (EtienneBerubeShopify) [inline lib/lhm/chunk_insert.rb:15]: Tagging the `insert ignore`. Might be seen as unnecessary, since if you're trying to insert on a replica it would throw an error regardless, but it would also make sense to tag "critical" queries. Any thoughts?

File: spec/test_helper.rb [hunk 14]
```diff
@@ -43,3 +47,20 @@ def printer.end(*) [] ;end
...
+    :username => db_config['master']['user'],
+    :password => db_config['master']['password'],
+    :port => db_config['master']['port']
+  )
+
+  conn.query("DROP DATABASE IF EXISTS #{$db_name}")
+  conn.query("CREATE DATABASE #{$db_name}")
+end
+
+init_test_db
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply spec/test_helper.rb:64]: From my understanding, `require` will only load a file and execute it if it has not been loaded yet. Meaning, this will only be loaded and executed once. At least this is my understanding of it, and there could be some edge cases when it comes to test environments where they could be scrapped after each test. [source](https://stackoverflow.com/questions/22757386/ruby-require-execute-multiple-times)

File: spec/test_helper.rb [hunk 15]
```diff
@@ -10,13 +10,17 @@
 require 'minitest/spec'
 require 'minitest/mock'
 require 'mocha/minitest'
+require 'method_hooks'
 require 'pathname'
+require 'byebug'
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply spec/test_helper.rb:15]: It is only for debugging. But it is in the `test_helper` so is not part of the actual gem. Can be removed, but might be re-added in the future by the next person to work on this gem.

File: spec/integration/sql_retry/retry_with_proxysql_spec.rb [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+    assert_equal e.class, Lhm::Error
+    assert_match(/LHM tried the reconnection procedure but failed. Latest error: Reconnected to wrong host/, e.message)
+
+    logs = @logger.string.split("\n")
+
+    assert logs.first.include?("Lost connection to MySQL, will retry to connect to same host")
+    assert logs.last.include?("Lost connection to MySQL server at 'reading initial communication packet")
+  end
+
+  it "Will abort if failover happens (mimicked with proxySQL)" do
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply spec/integration/sql_retry/retry_with_proxysql_spec.rb:88]: They are similar, yes. They take 2 approaches to ensure that the code behaves as intended. I'm not against having more tests than not, but I can understand that it could be seen as boilerplate and unnecessary. I kept it thinking it would not cause any harm, but it can be removed. What do you think?

File: spec/integration/sql_retry/retry_with_proxysql_spec.rb [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+    @lhm_retry = Lhm::SqlRetry.new(@connection, {},
+                                   true)
+  end
+
+  after(:each) do
+    # Restore default logger
+    Lhm.logger = @old_logger
+  end
+
+  it "Will abort if cannot find same host" do
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply spec/integration/sql_retry/retry_with_proxysql_spec.rb:31]: You're right, it could be clearer. I did mean to say permanently down, but the wording was lacking clarity. Will be changed.

File: docker-compose.yml [hunk 18]
```diff
@@ -44,3 +48,11 @@ services:
     command: "proxysql -c /etc/proxysql.cnf -f --idle-threads"
     ports:
       - "33005:3306"
+      - "6032:6032"
+  toxiproxy:
+    container_name: toxiproxy
+    image: "ghcr.io/shopify/toxiproxy"
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply docker-compose.yml:54]: It is indeed public. [source](https://github.com/Shopify/toxiproxy#1-installing-toxiproxy)

File: lib/lhm.rb (hunk 12 above)
Comment (EtienneBerubeShopify) [reply lib/lhm.rb:90]: fixed

File: README.md [hunk 19]
```diff
@@ -109,6 +109,48 @@ end
 to prevent accidental data loss. After successful or failed LHM migrations, these leftover
 tables must be cleaned up.
 
+### Usage with ProxySQL
+LHM has security features to recover from connection loss when used in conjunction with ProxySQL. There are multiple ways that
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply README.md:113]: fixed

File: lib/lhm.rb [hunk 20]
```diff
@@ -82,15 +83,26 @@ def cleanup_current_run(run, table_name, options = {})
...
+  #   Active / Deactivate ProxySQL-aware reconnection procedure (default to: false)
+  def setup(connection, connection_options = {})
+    @@connection = Connection.new(connection: connection, options: connection_options)
   end
 
-  def connection
+  # Setups DB connection
+  #
+  # @param [Hash] connection_options Optional options (defaults to: empty hash)
+  # @option options [Boolean] :reconnect_with_consistent_host
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply lib/lhm.rb:99]: Misread the documentation for `@param`. Fixed now. Good catch

File: lib/lhm.rb [hunk 21]
```diff
@@ -82,15 +83,26 @@ def cleanup_current_run(run, table_name, options = {})
...
+  #   Active / Deactivate ProxySQL-aware reconnection procedure (default to: false)
+  def connection(connection_options = {})
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply lib/lhm.rb:101]: I see what you mean, it can be dangerous. This is part of LHM interface and some gem users might be using it. There are 2 ways this could be addressed:
1) Change the way the method works, by only returning the instance of connection and raising an error if not defined (need to use `Lhm.setup` as a consistent way of setting a connection)
2) Find a way to hot-swap the new connection_options

Any preference?

File: lib/lhm/sql_retry.rb [hunk 22]
```diff
@@ -15,26 +16,128 @@ module Lhm
...
+      query = Lhm::ProxySQLHelper.tagged("SELECT #{name} LIMIT 1")
+
+      @connection&.execute(query).to_a.first.tap do |record|
+        return record&.first
+      end
+    end
+
+    def same_host_as_initial?
+      return @initial_server_id == server_id if cloudsql?
+      @initial_hostname == hostname
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply lib/lhm/sql_retry.rb:91]: During the initial investigation, I was told hostname would be a good way of ensuring uniqueness (at least at Shopify). I don't feel confident changing this so late, as all the initial testing has been done with hostname since the beginning. 
It could be changed to `server_id` if necessary, but it would require some additional testing to ensure that using `server_id` would not break any of the current implementation.

[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply lib/lhm/sql_retry.rb:91]: The incorporation of `server_id` was used as a fallback when used with CloudSQL.

File: lib/lhm/sql_retry.rb [hunk 23]
```diff
@@ -15,26 +16,128 @@ module Lhm
...
       @connection = connection
+      @log_prefix = options.delete(:log_prefix)
       @global_retry_config = default_retry_config.dup.merge!(options)
+      if (@with_consistent_host = with_consistent_host)
+        @initial_hostname = hostname
+        @initial_server_id = server_id
+      end
     end
 
+    # Complete explanation of algorithm: https://github.com/Shopify/db-engineering/issues/98#issuecomment-934948590
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply lib/lhm/sql_retry.rb:41]: Fixed to point to this PR + added flow diagram for the algorithm.

File: lib/lhm/sql_retry.rb [hunk 24]
```diff
@@ -15,26 +16,128 @@ module Lhm
...
+    def server_id
+      mysql_single_value(MYSQL_VAR_NAMES[:server_id])
+    end
+
+    def cloudsql?
+      mysql_single_value(MYSQL_VAR_NAMES[:version_comment]).include?(CLOUDSQL_VERSION_COMMENT)
+    end
+
+    def mysql_single_value(name)
+      query = Lhm::ProxySQLHelper.tagged("SELECT #{name} LIMIT 1")
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply lib/lhm/sql_retry.rb:82]: @connection is an ActiveRecord connection to not create a loop, since `Lhm::Connection` calls `with_retries`. Therefore, they need to be tagged manually.

File: lib/lhm/sql_retry.rb [hunk 25]
```diff
@@ -15,26 +16,128 @@ module Lhm
...
+
+      Retriable.retriable(retry_config) do
+        if @with_consistent_host
+          raise Lhm::Error.new("Could not reconnected to initial MySQL host. Aborting to avoid data-loss") unless same_host_as_initial?
+        end
+
         yield(@connection)
+      rescue StandardError => e
+        # Not all errors should trigger a reconnect. Some errors such be raised and abort the LHM (such as reconnecting to the wrong host).
+        raise e unless error_can_trigger_reconnect?(e)
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply lib/lhm/sql_retry.rb:57]: Good idea! I do agree that using error messages can be flaky. I'll investigate and see if I can find anything

File: lib/lhm/sql_retry.rb [hunk 26]
```diff
@@ -15,26 +16,128 @@ module Lhm
...
     def log_with_prefix(message, level = :info)
       message.prepend("[#{@log_prefix}] ") if @log_prefix
-      Lhm.logger.send(level, message)
+      Lhm.logger.public_send(level, message)
+    end
+
+    def reconnect_with_host_check!
+      log_with_prefix("Lost connection to MySQL, will retry to connect to same host")
+      begin
+        Retriable.retriable(host_retry_config) do
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply lib/lhm/sql_retry.rb:102]: Yes it is in another `retriable` block. We did discuss it when pairing. It is also O(n)^2, but n is known and is the number of retries possible.

File: lib/lhm/sql_retry.rb [hunk 27]
```diff
@@ -15,26 +16,128 @@ module Lhm
...
+          # tries to reconnect. On failure will trigger a retry
+          @connection.reconnect!
+          new_host = hostname
+          if new_host == @initial_hostname
+            # This is not an actual error, but it needs to trigger the Retriable
+            # from #with_retries to execute the desired logic again
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply lib/lhm/sql_retry.rb:108]: I tried to make it a bit clearer, let me know what you think.

File: lib/lhm/sql_retry.rb [hunk 28]
```diff
@@ -15,26 +16,128 @@ module Lhm
...
+            # New Master --> abort LHM (reconnecting will not change anything)
+            raise Lhm::Error.new("Reconnected to wrong host. Started migration on: #{@initial_hostname} (server_id: #{@initial_server_id}), but reconnected to: #{new_host} (server_id: #{@initial_server_id}).")
+          end
+        end
+      rescue StandardError => e
+        # The parent Retriable.retriable is configured to retry if it encounters an error with the success message.
+        # Therefore, if the connection is re-established successfully AND the host is the same, LHM can retry the query
+        # that originally failed.
+        raise e if reconnect_successful?(e)
+        # If the connection was not successful, the parent retriable will raise "unregistered" errors.
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply lib/lhm/sql_retry.rb:120]: Fixed

File: lib/lhm/sql_retry.rb [hunk 29]
```diff
@@ -57,8 +165,29 @@ def default_retry_config
...
+      {
+        on: {
+          StandardError => [
+            /Lost connection to MySQL server at 'reading initial communication packet'/
+          ]
+        },
+        multiplier: 1, # each successive interval grows by this factor
+        base_interval: 0.25, # the initial interval in seconds between tries.
+        tries: 20, # Number of attempts to make at running your code block (includes initial attempt).
+        rand_factor: 0, # percentage to randomize the next retry interval time
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply lib/lhm/sql_retry.rb:187]: Coarsly tuned. Taken from the "default_retry_config", but the base_interval was changed to 0.25s, which means it will try to reconnect for 5 seconds. What do you think these numbers should be?

File: spec/integration/lhm_spec.rb [hunk 30]
```diff
@@ -581,5 +582,77 @@
...
+        Lhm.logger = Logger.new(@logs)
+      end
+
+      it " should not try to reconnect if reconnect_with_consistent_host is not provided" do
+        connect_master_toxic!(with_retry: false)
+
+        table_create(:users)
+        100.times { |n| execute("insert into users set reference = '#{ n }'") }
+
+        assert_raises ActiveRecord::StatementInvalid do
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply spec/integration/lhm_spec.rb:600]: When something a MySQL backend is down, it will return a StatementInvalid bundled with the underlying MySQL2 error.

File: spec/integration/chunk_insert_spec.rb [hunk 31]
```diff
@@ -11,7 +11,8 @@
       @destination = table_create(:destination)
       @migration = Lhm::Migration.new(@origin, @destination)
       execute("insert into origin set id = 1001")
-      @instance = Lhm::ChunkInsert.new(@migration, connection, 1001, 1001)
+      @connection = Lhm::Connection.new(connection: connection)
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply spec/integration/chunk_insert_spec.rb:14]: I agree, but for the purpose of a test it's not terrible.

File: lib/lhm.rb (hunk 21 above)
[the remark this answers is not in the document]
Comment (shuhaowu) [reply lib/lhm.rb:101]: Following the principle of least surprises, the first one seems like it's more surprising if i think about it now.

File: lib/lhm/sql_retry.rb (hunk 26 above)
[the remark this answers is not in the document]
Comment (shuhaowu) [reply lib/lhm/sql_retry.rb:102]: I thought the retry backoff is exponential, so the time is O(exp(t)^2)? Is that still okay?

File: .github/workflows/test.yml [hunk 32]
```diff
@@ -15,8 +15,8 @@ jobs:
       - name: Set up Ruby
         uses: ruby/setup-ruby@v1
         with:
-          ruby-version: 2.5.1
-          bundler-cache: true
+          ruby-version: 2.7.0
+          bundler-cache: false
```
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply .github/workflows/test.yml:19]: Had trouble with wrong gems in bundler due to caching.

File: lib/lhm/sql_retry.rb (hunk 22 above)
[the remark this answers is not in the document]
Comment (shuhaowu) [reply lib/lhm/sql_retry.rb:89]: Let's not change this in this PR, but we should address this in a follow up and use only server_id, for example.

File: spec/integration/toxiproxy_helper.rb [hunk 33]
```diff
@@ -0,0 +1,44 @@
...
+      end unless restart_after.nil?
+
+      Toxiproxy[target].disable
+
+      yield
+
+    ensure
+      thread.join
+
+      #  Sometimes it's flaky and will keep it disabled
```
[the remark this answers is not in the document]
Comment (shuhaowu) [reply spec/integration/toxiproxy_helper.rb]: Fair enough. I'm glad we caught it!

File: lib/lhm/sql_retry.rb (hunk 29 above)
[the remark this answers is not in the document]
Comment (EtienneBerubeShopify) [reply lib/lhm/sql_retry.rb:173]: But what do you think? Is 5s of retries good enough to overcome the most common network blips? Especially since this will go on top of ActiveRecord's own retry logic

File: lib/lhm.rb [hunk 34]
```diff
@@ -82,16 +83,29 @@ def cleanup_current_run(run, table_name, options = {})
...
+  #
+  # @param [Hash] connection_options Optional options (defaults to: empty hash)
+  # @option connection_options [Boolean] :reconnect_with_consistent_host
+  #   Active / Deactivate ProxySQL-aware reconnection procedure (default to: false)
+  def connection(connection_options = nil)
+    if @@connection.nil?
+      raise 'Please call Lhm.setup' unless defined?(ActiveRecord)
+      @@connection = Connection.new(connection: ActiveRecord::Base.connection, options: connection_options || {})
+    else
+      @@connection.options = connection_options unless connection_options.nil?
```
Comment (shuhaowu) [inline lib/lhm.rb:106]: Wait, I thought the idea of this method is to just return @@connection if it exist and error if it doesn't? This kind of overwrite feels like it might be fragile.

File: lib/lhm/connection.rb [hunk 35]
```diff
@@ -8,36 +9,74 @@ class Connection < SimpleDelegator
...
 
       # Creates delegation for the ActiveRecord Connection
       super(connection)
     end
 
-    def execute(query, retry_options = {})
-      exec_with_retries(:execute, query, retry_options)
+    def options=(options)
+      # If any other flags are added. Add the "processing" here
+      @sql_retry.reconnect_with_consistent_host = options[:reconnect_with_consistent_host] || false
```
Comment (shuhaowu) [inline lib/lhm/connection.rb:27]: So if we do `connection.options = options`, it will only update a single member variable in `@sql_retry`? Why not just do `connections.reconnect_with_consistent_host = true/false` instead of using a options hash?

File: lib/lhm/connection.rb [hunk 36]
```diff
@@ -8,36 +9,74 @@ class Connection < SimpleDelegator
...
+        exec(:select_values, query)
+      end
+    end
+
+    def select_one(query, should_retry: false, retry_options: {})
+      if should_retry
+        exec_with_retries(:select_one, query, retry_options)
+      else
+        exec(:select_one, query)
+      end
```
Comment (shuhaowu) [inline lib/lhm/connection.rb:67]: Fine for now, it occured to me that AR allows for more than just queries to be passed in here (and possibly also bind arguments?), not that we use this feature..

File: lib/lhm/connection.rb [hunk 37]
```diff
@@ -8,36 +9,53 @@ class Connection < SimpleDelegator
...
+        exec_with_retries(:select_value, query, retry_options)
+      else
+        exec(:select_value, query)
+      end
     end
 
     private
 
+    def exec(method, sql)
+      connection.public_send(method, Lhm::ProxySQLHelper.tagged(sql))
```
[the remark this answers is not in the document]
Comment (shuhaowu) [reply lib/lhm/connection.rb:73]: For the QueryLog, what I mean is if that is set, will the query be double tagged? Is that OK?

For `method_missing`, we don't need it now, but as we override more and more methods, we might consider using it. Let's defer that for the future tho as maybe at some point we can switch over to mysql2 instead of the AR connection.

**Review by shuhaowu — APPROVED** [review]
Let's NOT merge this for now due to infrastructure freeze as I want to keep the master branch to the latest version, in case we have to do any emergency fixes. If we merge this to master, doing emergency fixes might be more challenging.

File: lib/lhm/connection.rb (hunk 35 above)
Comment (EtienneBerubeShopify) [reply lib/lhm/connection.rb:27]: I did because options could contain more than one values. I also added this so that there could be more feature flags in the future. However, if you don't see any new features coming in the future,  I can always make it more clear.

File: lib/lhm.rb (hunk 34 above)
Comment (EtienneBerubeShopify) [reply lib/lhm.rb:106]: This is also what I thought, but there was an assignment in there. I didn't want to break the "developer facing" interface, so I kept it as is. 

Original code:
```ruby
  def connection
    @@connection ||=
      begin
        raise 'Please call Lhm.setup' unless defined?(ActiveRecord)
        ActiveRecord::Base.connection
      end
  end
```

The decision came from this thread that we had. [[source](https://github.com/Shopify/lhm/pull/112#discussion_r746954532)]
